client_credentials_auth.go

```diff
@@ -0,0 +1,104 @@
+package sdk
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+)
+
+type ClientCredentialsTokenSource struct {
+	clientID     string
+	clientSecret string
+	tokenURL     string
+	scope        string
+	grantType    string
+	token        *ClientCredentialsToken
+}
+
+type ClientCredentialsAuth struct {
+	tokenSource *ClientCredentialsTokenSource
+}
+
+func (cca *ClientCredentialsAuth) Set(req *http.Request) error {
+	token, err := cca.tokenSource.Token()
+	if err != nil {
+		return err
+	}
+	req.Header.Add("Authorization", "Bearer "+token)
+	return nil
+}
+
+func NewClientCredentialsTokenSource(clientID, clientSecret, tokenURL, scope, grantType string) *ClientCredentialsTokenSource {
+	return &ClientCredentialsTokenSource{
+		clientID:     clientID,
+		clientSecret: clientSecret,
+		tokenURL:     tokenURL,
+		scope:        scope,
+		grantType:    grantType,
+	}
+}
+
+func (ts *ClientCredentialsTokenSource) Token() (string, error) {
+
+	if ts.token == nil || ts.token.Expired() {
+
+		reqBody := url.Values{}
+		reqBody.Set("client_id", ts.clientID)
+		reqBody.Set("client_secret", ts.clientSecret)
+		reqBody.Set("grant_type", ts.grantType)
+		reqBody.Set("scope", ts.scope)
+
+		buffer := []byte(reqBody.Encode())
+
+		req, err := http.NewRequest(http.MethodPost, ts.tokenURL, bytes.NewBuffer(buffer))
+		if err != nil {
+			return "", err
+		}
+		req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+		res, err := http.DefaultClient.Do(req)
+		if err != nil {
+			return "", err
+		}
+
+		var body []byte
+		if res.Body != nil {
+			defer res.Body.Close()
+			body, _ = io.ReadAll(res.Body)
+		}
+
+		if res.StatusCode != http.StatusOK {
+			return "", fmt.Errorf("unexpected status code %d, body: %s", res.StatusCode, string(body))
+		}
+
+		token := &ClientCredentialsToken{}
+		if err := json.Unmarshal(body, token); err != nil {
+			return "", fmt.Errorf("unable to unmarshal token: %s", err)
+		}
+
+		token.ObtainedAt = time.Now()
+
+		ts.token = token
+
+		return ts.token.AccessToken, nil
+	}
+	return ts.token.AccessToken, nil
+}
+
+type ClientCredentialsToken struct {
+	AccessToken string `json:"access_token"`
+	TokenType   string `json:"token_type"`
+	ExpiresIn   int    `json:"expires_in"`
+	ObtainedAt  time.Time
+}
+
+func (t *ClientCredentialsToken) Expired() bool {
+	if t.ExpiresIn == 0 {
+		return false
+	}
+	return t.ObtainedAt.Add(time.Duration(t.ExpiresIn) * time.Second).Before(time.Now())
```

Comment:
I think we should take into account a small delta to avoid late expirations due to client-server time mismatches.

Like we did in `token.go`: https://github.com/openfaas/go-sdk/blob/73f84ff5d81d87ba6d1ea5c8672ebb72600966d3/token.go#L13-L16